src/ffi/logging.rs

```diff
@@ -0,0 +1,236 @@
+use std::{
+    io::{self},
+    str::from_utf8,
+    sync::{Arc, Mutex},
+};
+
+use tracing::{level_filters::LevelFilter, Subscriber};
+use tracing_subscriber::{
+    field::delimited::Delimited,
+    fmt::{self, format::debug_fn, FormatEvent, FormatFields, MakeWriter},
+    layer::SubscriberExt,
+    registry::LookupSpan,
+};
+
+use crate::{TelioLogLevel, TelioLoggerCb};
+
+pub fn build(log_level: crate::TelioLogLevel, logger: Box<dyn TelioLoggerCb>) -> impl Subscriber {
+    let field_fmt = Delimited::new(
+        " ",
+        debug_fn(|writer, field, value| {
+            if field.name() == "message" {
+                write!(writer, "{value:?}")
+            } else {
+                write!(writer, "{field}={value:?}")
+            }
+        }),
+    );
+
+    tracing_subscriber::registry()
+        .with(LevelFilter::from_level(log_level.into()))
+        .with(
+            fmt::layer()
+                .event_format(TelioEventFmt)
+                .fmt_fields(field_fmt)
+                .with_writer(FfiCallback::new(logger)),
+        )
+}
+
+struct TelioEventFmt;
+
+impl<S, N> FormatEvent<S, N> for TelioEventFmt
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+    N: for<'a> FormatFields<'a> + 'static,
+{
+    fn format_event(
+        &self,
+        ctx: &fmt::FmtContext<'_, S, N>,
+        mut writer: fmt::format::Writer<'_>,
+        event: &tracing::Event<'_>,
+    ) -> std::fmt::Result {
+        let meta = event.metadata();
+        write!(
+            writer,
+            "{:?}:{} ",
+            meta.module_path().unwrap_or("<unknown module>"),
+            meta.line().unwrap_or(0),
+        )?;
+
+        ctx.format_fields(writer.by_ref(), event)?;
+
+        writeln!(writer)
+    }
+}
+
+pub struct FfiCallback {
+    callback: Arc<dyn TelioLoggerCb>,
+}
+impl FfiCallback {
+    fn new(logger: Box<dyn TelioLoggerCb>) -> Self {
+        Self {
+            callback: logger.into(),
+        }
+    }
+}
+
+impl<'a> MakeWriter<'a> for FfiCallback {
+    type Writer = FfiCallbackWriter;
+
+    fn make_writer(&self) -> Self::Writer {
+        unreachable!("make_writer_for is used")
```

Comment:
This message sounds weird in this context imho,  maybe something like "make_writer is not expected to be used"?

Reply:
fixed

Reply:
done